ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/ProjectOperator.java

```diff
@@ -0,0 +1,227 @@
+  private ProjectNode logicalNode;
+  private TableRow row;
+  private KsqlTransformer<Object, GenericRow> transformer;
+  private Function<TableRow, GenericRow> preSelectTransform;
+
+  public ProjectOperator(
+      final KsqlConfig ksqlConfig,
+      final MetaStore metaStore,
+      final Materialization mat,
+      final ImmutableAnalysis analysis,
+      final KsqlExecutionContext executionContext,
+      final Stacker contextStacker,
+      final ProjectNode logicalNode,
+      final LogicalSchema outputSchema,
+      final boolean isSelectStar
+
+  ) {
+    this.ksqlConfig = Objects.requireNonNull(ksqlConfig, "config");
+    this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+    this.mat = Objects.requireNonNull(mat, "mat");
+    this.analysis = Objects.requireNonNull(analysis, "analysis");
+    this.executionContext = Objects.requireNonNull(executionContext, "executionContext");
+    this.contextStacker = Objects.requireNonNull(contextStacker, "contextStacker");
+    this.logicalNode = Objects.requireNonNull(logicalNode, "logicalNode");
+    this.outputSchema = Objects.requireNonNull(outputSchema, "outputSchema");
+    this.isSelectStar = isSelectStar;
+  }
+
+  @Override
+  public void open() {
+    child.open();
+    if (isSelectStar) {
+      return;
+    }
+
+    final LogicalSchema inputSchema = mat.schema();
+
+    final boolean noSystemColumns = analysis.getSelectColumnNames().stream()
+        .noneMatch(SystemColumns::isSystemColumn);
+    final boolean noKeyColumns = analysis.getSelectColumnNames().stream()
+        .noneMatch(inputSchema::isKeyColumn);
+
+    final LogicalSchema intermediateSchema;
+    if (noSystemColumns && noKeyColumns) {
+      intermediateSchema = inputSchema;
+      preSelectTransform = TableRow::value;
+    } else {
+      // SelectValueMapper requires the rowTime & key fields in the value schema :(
+      final boolean windowed = mat.windowType().isPresent();
+
+      intermediateSchema = inputSchema
+          .withPseudoAndKeyColsInValue(windowed);
+
+      preSelectTransform = row -> {
+        final Struct key = row.key();
+        final GenericRow value = row.value();
+
+        final List<Object> keyFields = key.schema().fields().stream()
+            .map(key::get)
+            .collect(Collectors.toList());
+
+        value.ensureAdditionalCapacity(
+            1 // ROWTIME
+            + keyFields.size()
+            + row.window().map(w -> 2).orElse(0)
+        );
+
+        value.append(row.rowTime());
+        value.appendAll(keyFields);
+
+        row.window().ifPresent(window -> {
+          value.append(window.start().toEpochMilli());
+          value.append(window.end().toEpochMilli());
+        });
+
+        return value;
+      };
+    }
+
+    final SelectValueMapper<Object> select = SelectValueMapperFactory.create(
+        logicalNode.getSelectExpressions(),
+        intermediateSchema,
+        ksqlConfig,
+        metaStore
+    );
+
```

Comment:
Does anyone know what the code below is doing and why it is needed?

Reply:
I think because `SelectValueMapper` is shared with persistent queries and it expects this so it can write errors to the processing topic.  I'm not sure if we really would want that for pull queries or if it's just shoehorned in here because the code is shared.